<html>
<meta http-equiv="X-UA-Compatible" content="IE=10" />
<body>
<script>
cAsserts = 33;
msAssertCheck = 2000;
</script>
<script src="../test.js"></script>
<div id="output"></div>
<script>
assert = function( sMsg, exp, got ) {
	Test.assert( got, exp, sMsg );
}
// Async
window.name = "doing";
msw1 = 40;
msw2 = 20;
waiter = function( args ) {
	var async = __.Async.promise( args );
	var ms = Math.round( Math.random() * msw1 );
	setTimeout( function() {
		var aResult = args && args.aResult;
		if( ! aResult ) {
			aResult = []
		}
		var s = ( args && args.x ) ? args.x : "_";
		if( s == "error" ) {
			async.reject( { sError: "an error happend" } );
		}
		else {
			aResult.push( "w" + s );
			async.resolve( { aResult : aResult} );
		}
	}, ms );
}
waiter2 = function( args ) {
	var async = __.Async.promise( args );
	var ms = Math.round( Math.random() * msw2 );
	setTimeout( function() {
		var aResult = args && args.aResult;
		if( ! aResult ) {
			aResult = []
		}
		var s = ( args && args.x ) ? args.x : "_";
		aResult.push( "w2" + s );
		async.resolve( { aResult : aResult} );
	}, ms );
}
var error = function( args ) {
	var async = __.Async.promise( args );
	var ms = args.ms || 1;
	setTimeout( function() {
		var s = "[" + args.s + "]";
		//console.log( s, args.sRes );
		args.sRes = ( args.sRes ) ? args.sRes + s : s;
		if( args.bErr ) {
			async.reject( { sError : s } );
		}
		else {
			async.resolve( { sOk : s } );
		}
	}, ms );
}
var error2 = function( args ) {
	var async = __.Async.promise( args );
	setTimeout( function() {
		var s = "[" + args.s + "]";
		args.sRes = ( args.sRes ) ? args.sRes + s : s;
		if( args.bErr ) {
			async.reject( { sError : s } );
		}
		else {
			async.resolve( { sOk : s } );
		}
	}, 1 );
}

para = function( args ) {
	var async = __.Async.promise( args );
	setTimeout( function() {
		var aResult = args && args.aResult;
		if( ! aResult ) {
			aResult = []
		}
		var s = ( args && args.x ) ? args.x : "_";
		aResult.push( "pp" + s );
		async.resolve( { aResult : aResult} );
	}, 100 );
}
para2 = function( args ) {
	var async = __.Async.promise( args );
	setTimeout( function() {
		var paraResult = args && args.paraResult;
		if( ! paraResult ) {
			paraResult = []
		}
		var s = ( args && args.x ) ? args.x : "_";
		paraResult.push( "p" + s );
		async.resolve( { paraResult : paraResult} );
	}, 50 );
}
fnStatus=function( args ){}
// error async
Error = new __.Async({
	  sdftError : "hi mom"
	, id : "this.namespace.Error"
	, fnerr : function( oError ) {
		assert( "error message", 'id:(this.namespace.Error) task:(error2) args:({"__guid_async__":"Async1","x":"error","aResult":["w1"]}) sdftError:(hi mom) sError:("an error happend") ', oError.sStack );
	}
})
.then( this, "waiter", { x : 1 }, "error1" )
.then( this, "waiter", { x : "error" }, "error2" )
.then( this, "waiter", { x : 1 }, "error3" )
.then( this, "waiter", { x : 1 }, "error4" )
.start();

// error async
Error = new __.Async({
	  id : "test error"
	, sFile : "test.async.html:290"
	, fnerr : function( oError ) {
		assert( "error message 2", 'id:(test error) task:(error2) args:({"__guid_async__":"Async2","x":"error","aResult":["w1"]}) sdftError:() sError:("an error happend") ', oError.sStack );
	}
})
.then( this, "waiter", { x : 1 }, "error1" )
.then( this, "waiter", { x : "error" }, "error2" )
.then( this, "waiter", { x : 1 }, "error3" )
.then( this, "waiter", { x : 1 }, "error4" )
.start();
o = {
	do : function() {
		var that = this;
		( new __.Async() )
		.then( that, "getRecords", { idUser : 123 }, "Getting records" )
		.then( function( args ) {
			__.Async.promise( args ).resolve();
		}, "Printing records" )
		.start();
	}
	, getRecords : function( args ) {
		setTimeout( function() {
			var hResult = "<b>no records</b>";
			__.Async.promise( args ).resolve( { hResult: hResult } );
		}, 2000 ); 
	}
}
o.do()

var a1 = new __.Async( { fnstat : fnStatus } )
.then( this, "waiter", { x : 4 }, "a1.1" )
.then( this, "waiter", "a1.2" )
.wait( function() {
	return window.name == "done";
}, 800, "wait for window name" )
.wait( function() {
	return window.name == "done";
} )
.then( this, "waiter", { x : 6 }, "a1.3" )
.then( this, "waiter", { x : 2 }, "a1.4" )
.then( function( args ) {
	var async = __.Async.promise( args );
	[1,2,3].forEach( function( n ) {
		async.then( this, "waiter", { x : n }, "a1.5." + n )
	} );
	async.resolve();
}, "a1.5" )
.then( this, "waiter", { x : 6 }, "a1.6" )
.then( this, "waiter", { x : 2 }, "a1.7" )
.then( function( args ) {
	assert( "wait and creation in loop", ["w4","w4","w6","w2","w1","w2","w3","w6","w2"], args.aResult );
	__.Async.promise( args ).resolve();
}, "last msg" )
.start();


var a2 = new __.Async()
.then( this, "waiter2", "a2.1" )
.then( this, "waiter2", "a2.2" )
.then( this, "waiter2", { x : 20 }, "a2.3" )
.then( this, "waiter2", "a2.4" )
.then( this, "waiter2", "a2.4" )
.pause( 400 )
.then( function( args ) {
	window.name = "done";
	__.Async.promise( args ).resolve();
}, "set window name" )
.then( this, "waiter2", "a2.4" )
.then( function( args ) {
	assert( "simple with wait condition", ["w2_","w2_","w220","w220","w220","w220"], args.aResult );
	__.Async.promise( args ).resolve();
}, "first last statement" )
.clear()
.then( this, "waiter2", "a2.5" )
.then( function( args ) {
	assert( "anonymous function as last task", [ "w2_" ], args.aResult );
	__.Async.promise( args ).resolve();
}, "last statement" )
.start()

var mixed = new __.Async()
.then( this, "waiter", "mixed.1"  )
.clear()
.then( this, "waiter", "mixed.1"  )
.then( this, "waiter", "mixed.1"  )
.then( this, "waiter2", "mixed.2"  )
.then( function( args ) {
	var async = __.Async.promise( args );
	async.then( this, "waiter2", "mixed.3"  )
	async.resolve();
} )
.then( this, "waiter", "mixed.4"  )
.then( function( args ) {
	assert( "mixed functions", ["w_","w_","w2_","w2_","w_"], args.aResult );
	__.Async.promise( args ).resolve();
} )
.start()

//  test with objects
obj = {
	  owaiter : function( args ) {
		var async = __.Async.promise( args );
		var ms = Math.round( Math.random() * 300 );
		setTimeout( function() {
			var aResult = args && args.aResult;
			if( ! aResult ) {
				aResult = []
			}
			var s = ( args && args.x ) ? args.x : "_";
			aResult.push( "ow" + s );
			async.resolve( { aResult : aResult} );
		}, ms );
	}
	, owaiter2 : function( args ) {
		var async = __.Async.promise( args );
		var ms = Math.round( Math.random() * 10 );
		setTimeout( function() {
			var aResult = args && args.aResult;
			if( ! aResult ) {
				aResult = []
			}
			var s = ( args && args.x ) ? args.x : "_";
			aResult.push( "ow2" + s );
			async.resolve( { aResult : aResult} );
		}, ms );
	}
	, first : function( cbfn ) {
		var that = this;
		( new __.Async() )
		.then( this, "owaiter2", "ofirst.1"  )	
		.then( this, "owaiter2", "ofirst.2"  )	
		.then( function( args ) {
			var async = __.Async.promise( args );
			async.then( that, "owaiter2", "ofirst.3"  );
			async.resolve();
		} )
		.then( function( args ) {
			assert( "object nested", ["ow2_","ow2_","ow2_"], args.aResult );
			__.Async.promise( args ).resolve();
			cbfn();
		} )
		.start();
	}
	, init : function() {
		var that = this;
		this.first( function() {
			( new __.Async() )
			.then( that, "owaiter", "oinit.1"  )	
			.then( that, "owaiter", "oinit.2" )	
			.then( that, "owaiter", "oinit.3" )	
			.then( function( args ) {
				assert( "inner object nested", ["ow_","ow_","ow_"], args.aResult );
				__.Async.promise( args ).resolve();
			} )
			.start();
		} );	
	}
};
obj.init();

// deeply nested
var laster = function( args ) {
	var async = __.Async.promise( args );
	async.then( function( args ) {
		var async = __.Async.promise( args );
		async.then( self, "waiter", { x : 341 }, "LASTER1" )
		async.then( self, "waiter", { x : 342 }, "LASTER1" )
		async.then( self, "waiter", { x : 343 }, "LASTER1" )
		async.resolve();	
	}, "LASTER0" )
	async.then( self, "waiter", { x : 4 }, "LASTER1" )
	async.then( self, "waiter", { x : 5 }, "LASTER2" )
	async.resolve();
}
var taxer = function( args ) {
	var async = __.Async.promise( args );
	async.then( self, "waiter", { x : 3 }, "TAXER1" )
	async.then( self, "laster", "TAXER -> LAST" )
	async.then( self, "waiter", { x : 6 }, "TAXER2" )
	async.resolve();
}
var cache = function( args ) {
	var async = __.Async.promise( args );
	async.then( self, "waiter", { x : 2 }, "CACHE7" )
	async.then( self, "taxer", "CACHE -> TAX" )
	async.then( self, "waiter", { x : 7 }, "CACHE77" )
	async.then( self, "waiter", { x : 8 }, "CACHE777" )
	async.resolve();
};
Schacht = new __.Async();
Schacht.then( this, "waiter", { x : 1 } ,"schacht1" )
.then( this, "cache", "schacht3" )
.then( this, "waiter", { x : 9 }, "schacht4" )
.then( function( args ) {
	assert( "deeply nested", ["w1","w2","w3","w341","w342","w343","w4","w5","w6","w7","w8","w9"], args.aResult );
	__.Async.promise( args ).resolve();
} )
.start();

// last late arrival 
LastLate = new __.Async()
.then( this, "waiter", { x : 1 } ,"ll1" )
.then( this, "waiter", { x : 2 }, "ll2" )
.then( function( args ) {
	var async = __.Async.promise( args )
	async.then( self, "waiter", { x : 3 }, "ll3.1" )
	async.then( self, "waiter", { x : 4 }, "ll3.2" )
	async.then( self, "waiter", { x : 5 }, "ll3.3" )
	async.then( self, "waiter", { x : 6 }, "ll3.4" )
	async.resolve();
}, "ll3" )
.then( this, "waiter", { x : 7 } ,"ll3" )
.then( this, "waiter", { x : 8 }, "ll4" )
.then( function( args ) {
	assert( "late arrivals", ["w1", "w2", "w3", "w4", "w5", "w6", "w7", "w8"], args.aResult );
	__.Async.promise( args ).resolve();
} )
.start();

// stop async
Stop = new __.Async()
.then( this, "waiter", { x : 1 } ,"stop1" )
.then( function( args ) {
	__.Async.promise( args ).stop();
	setTimeout( function() {
		assert( "stop tasks", [ "w1" ], args.aResult );
		
	}, 1000 );
}, "stop2" )
.then( this, "waiter", { x : 1 } ,"stop3" )
.then( this, "waiter", { x : 1 } ,"stop4" )
.start();

// async call w/o promise

var noprom = function( args ) {
	var async = __.Async.promise( args );
	async.then( function( args ) {
		var async = __.Async.promise( args );
		async.then( self, "waiter", { x : 341 }, "LASTER1" )
		async.then( self, "waiter", { x : 342 }, "LASTER2" )
		async.then( self, "waiter", { x : 343 }, "LASTER3" )
		async.resolve();	
	}, "LASTER0" )
	async.then( self, "waiter", { x : 4 }, "LASTER4" )
	async.then( self, "waiter", { x : 5 }, "LASTER2" )
	async.then( function( args ) {
		__.Async.promise( args ).resolve();
		assert( "no promise object", ["w341","w342","w343","w4","w5"], args.aResult );
	} );
	async.resolve();
}
noprom();

( new __.Async() )
.then( "noprom" )
.then( function( args ) {
	__.Async.promise( args ).resolve();
	assert( "no promise object with promise", ["w341","w342","w343","w4","w5"], args.aResult );
} )
.start();
// test pass on args
var argstest = function( args ) {
	var async = __.Async.promise( args );
	args.aResult.push( "|" + args.y + "|" );
	if( ! args.laTest ) {
		args.laTest = [];
	}
	args.laTest.push( { "a" : [ args.y, "@" + args.y ] } )
	async.resolve()
};
( new __.Async() )
.then( self, "waiter", { x : 1}, "args1" )
.then( self, "waiter", { x : 2 }, "args2" )
.then( self, "waiter", { x : 3 }, "args3" )
.then( function( args ) {
	__.Async.promise( args ).resolve();

}, "args4" )
.then( self, "argstest", { y : { arg : "x" } }, "args5!!" )
.then( self, "waiter", { x : 4}, "args6" )
.then( self, "waiter", { x : 5 }, "args7" )
.then( self, "argstest", { y : { arg : "laTest.0.a.1" } }, "args8!!" )
.then( function( args ) {
	__.Async.promise( args ).resolve();
	assert( "late come args 1", ["w1","w2","w3","|3|","w4","w5","|@3|"], args.aResult );
	assert( "late come args 2", [ {"a":[3, "@3"]},{"a":["@3","@@3"]}], args.laTest );
} )
.start();

// then arguments handling
func = function( args ) {
	var async = __.Async.promise( args );
	var sMsg = args.sMsg || "na";
	async.resolve( args );
};
( new __.Async() )
.then( self, "func", { sMsg : "one" }, "args2_1" )
.then( self, "func", "args2_2" )
.then( self, "func", { sMsg : "three" }, "args2_3" )
.then( function( args ) {
	__.Async.promise( args ).resolve();
	assert( "log message", {__guid_async__: "Async14", sMsg: "three"}, args );
}, "args2_4" )
.start();

// test standalone
getKey = function( args ) {
	var async = __.Async.promise( args );
	setTimeout( function() {
		var async = __.Async.promise( args );
		async.resolve( { "key" : "$ecretke!" + args.salt } );
	}, 100 );
	//async.resolve();
};

( new __.Async() )
.then( self, "getKey", { salt : "X" }, "key_1" )
.then( function( args ) {
	__.Async.promise( args ).resolve();
	assert( "standalone_in_async", "$ecretke!X", args.key );
}, "key_result" )
.start();

getKey( { salt : "Y"
	, cb : function( args ) {
		assert( "standalone", "$ecretke!Y", args.key );
	}
} );
getKey( { salt : "Z" } );
// parallel async tasks
var parallel = new __.Async()
.then( this, "waiter", {x:1}, "p1" )
.then( this, "waiter", {x:2}, "p2" )
.then( this, "waiter", {x:3}, "p3" )
.when( [
	  ( new __.Async() )
		.then( this, "para", {x:21}, "p21" )
		.then( this, "para", {x:22}, "p22" )
		.then( this, "para", {x:23}, "p23" )
		.then( function( args ) {
			var async = __.Async.promise( args );
			var p1Results = args.aResult;
			delete args.aResult;
			async.resolve( { p1Results : p1Results } );
		} )
	 , ( new __.Async() )
		.then( this, "para", {x:32}, "p32" )
		.then( this, "para", {x:33}, "p33" )
		.then( this, "para", {x:34}, "p34" )
		.then( function( args ) {
			var async = __.Async.promise( args );
			var p2Results = args.aResult;
			delete args.aResult;
			async.resolve( { p2Results : p2Results } );
			
		} )
	, ( new __.Async() ).then( this, "para2", {x:42}, "p42" )
	, ( new __.Async() ).then( this, "para2", {x:52}, "p52" )
	, ( new __.Async() ).then( this, "para2", {x:62}, "p62" )

], "execute in parallel" )
.then( this, "waiter", {x:4}, "p4" )
.then( this, "waiter", {x:5}, "p5" )
.then( this, "waiter", {x:6}, "p6" )
.then( this, "waiter", {x:7}, "p7" )
.then( function( args ) {
	assert( "parallel1", ["w1", "w2", "w3", "w4", "w5", "w6", "w7"], args.aResult );
	assert( "parallel2",["pp21", "pp22", "pp23"]
 , args.p1Results );
	assert( "parallel3",["pp32", "pp33", "pp34"] , args.p2Results );
	assert( "parallel4",["p62"]
 , args.paraResult );
	__.Async.promise( args ).resolve();
} )
.start()

// try catch

// simple try catch
var err = ( new __.Async( {
	fnerr : function( oError ) {
		console.log( ">>>>>>>>>", oError );
	}
} ) )
.then( this, "error", {bErr:0,s:"good"} )
.try()
.then( this, "error", {bErr:1,s:"bad"} )
.catch()
.then( this, "error", {bErr:0,s:"g00d"} )
.then( function( args ) {
	Test.assert( args.sRes, "[good][bad][g00d]", "simple try catch, one task" );
} )
.start();


var err = ( new __.Async( {
	fnerr : function( oError ) {
		//c_onsole.log( ">>>>>>>>>", oError );
	}
} ) )
.then( this, "error", {bErr:false,s:"good"} )
.try()
	.then( this, "error", {bErr:true,s:"bad1"} )
	.then( this, "error", {bErr:true,s:"bad2"} )
	.then( this, "error", {bErr:true,s:"bad3"} )
.catch()
	.then( this, "error", {bErr:false,s:"g00d"} )
.then( function( args ) {
	Test.assert( args.sRes, "[good][bad1][g00d]", "simple try catch, many tasks" );
} )
.start();

var err = ( new __.Async( {
	fnerr : function( oError ) {
		//c_onsole.log( ">>>>>>>>>", oError );
	}
} ) )
.then( this, "error", {bErr:false,s:"good"} )
.try().then( this, "error", {bErr:true,s:"bad1"} ).catch()
.try().then( this, "error", {bErr:true,s:"bad2"} ).catch()
.try()
	.then( this, "error", {bErr:true,s:"bad3"} )
	.then( this, "error", {bErr:true,s:"bad4"} )
.catch()
.then( this, "error", {bErr:false,s:"g00d"} )
.then( function( args ) {
	Test.assert( args.sRes, "[good][bad1][bad2][bad3][g00d]", "multiple catches" );
} )
.start();


var err = ( new __.Async( {
	fnerr : function( oError ) {
		//c_onsole.log( ">>>>>>>>>", oError );
	}
} ) )
err.try()
.then( this, "error", {bErr:true,s:"bad"} )
.catch( function( args ) {
	var async = __.Async.promise( args );
	async.then( this, "error2", {bErr:false,s:"DMIEP"} )
	async.then( this, "error2", {bErr:false,s:"AWIEGL"} )
	async.resolve();
} )
.then( this, "error", {bErr:false,s:"good2"} )
.then( function( args ) {
	Test.assert( args.sRes, "[bad][DMIEP][AWIEGL][good2]", "try catch with func" );
} )
.start();

var err = ( new __.Async( {
	fnerr : function( oError ) {
		//c_onsole.log( ">>>>>>>>>", oError );
	}
} ) )
err.try()
.then( this, "error", {bErr:false,s:"notsobad"} )
.catch( function( args ) {
	var async = __.Async.promise( args );
	async.then( this, "error2", {bErr:false,s:"DMIEP"} )
	async.then( this, "error2", {bErr:false,s:"AWIEGL"} )
	async.resolve();
} )
.then( this, "error", {bErr:false,s:"good2"} )
.then( function( args ) {
	Test.assert( args.sRes, "[notsobad][good2]", "try catch with func and no error" );
} )
.start();

var err = ( new __.Async() );
err.try()
.then( this, "error", {bErr:true,s:"bad"} )
.catch( function( args ) {
	var async = __.Async.promise( args );
	if( args.bErr ) {
		async.then( this, "error2", {bErr:false,s:"DMIEP"} )
	}
	async.resolve();
} )
.then( this, "error", {bErr:false,s:"good2"} )
.then( function( args ) {
	Test.assert( args.sRes, "[bad][DMIEP][good2]", "try catch with func and no error" );
} )
.start();
// parallel tasks with try catch
app = {
	test : function() {
		var lsRes = [];
		var err = ( new __.Async() );
		err.then( self, "error", {bErr:false,s:"good1"}, "ulla" )
		.when( [
			  ( new __.Async() )
				.try()
				.then( self, "error", {bErr:false,s:"good1"}, "g1" )
				.then( function( args ) {
					var async = __.Async.promise( args );
					lsRes.push( args.s );
					async.resolve();
				}, "g2" )
				.catch()
			, ( new __.Async() )
				.try()
				.then( self, "error", {bErr:true,s:"bad1"}, "b1" )
				.then( function( args ) {
					var async = __.Async.promise( args );
					lsRes.push( args.s );
					async.resolve();
				}, "b2" )
				.catch()
			, ( new __.Async() )
				.try()
				.then( self, "error", {bErr:false,s:"good2"} )
				.then( function( args ) {
					var async = __.Async.promise( args );
					lsRes.push( args.s );
					async.resolve();
				} )
				.catch()
			, ( new __.Async() )
				.try()
				.then( self, "error", {bErr:true,s:"bad2"} )
				.then( function( args ) {
					var async = __.Async.promise( args );
					lsRes.push( args.s );
					async.resolve();
				} )
				.catch()
		] )
		.then( function( args ) {
			var async = __.Async.promise( args );
			Test.assert( lsRes, ["good1","good2"], "try catch with parallel tasks" );
			async.resolve();
		} )
		.start()
	}
}
app.test();
app2 = {
	test : function() {
		var lsRes = [];
		var err = ( new __.Async() );
		err.try()
		.when( [
			  ( new __.Async() )
				.then( self, "error", {bErr:false,ms:10,s:"good1"}, "g1" )
				.then( function( args ) {
					var async = __.Async.promise( args );
					lsRes.push( args.s );
					async.resolve();
				}, "g2" )
			, ( new __.Async() )
				.then( self, "error", {bErr:true,ms:100,s:"bad1"}, "b1" )
				.then( function( args ) {
					var async = __.Async.promise( args );
					lsRes.push( args.s );
					async.resolve();
				}, "b2" )
			, ( new __.Async() )
				.then( self, "error", {bErr:false,s:"good2"} )
				.then( function( args ) {
					var async = __.Async.promise( args );
					lsRes.push( args.s );
					async.resolve();
				} )
			, ( new __.Async() )
				.then( self, "error", {bErr:true,ms:100,s:"bad2"} )
				.then( function( args ) {
					var async = __.Async.promise( args );
					lsRes.push( args.s );
					async.resolve();
				} )
		] )
		.catch( function( args ) {
			var async = __.Async.promise( args );
			lsRes.push( "catch" );
			async.resolve();

		} )
		.then( function( args ) {
			var async = __.Async.promise( args );
			Test.assert( lsRes, ["good2","good1","catch"], "try catch with parallel tasks wrapping" );
			async.resolve();
		} )
		.start()
	}
}
app2.test();



app3 = {
	test : function() {
		var lsRes = [];
		var err = ( new __.Async() );
		err.try()
		.when( [
			  ( new __.Async() )
				.then( self, "error", {bErr:false,ms:100,s:"good1"}, "g1" )
				.then( function( args ) {
					var async = __.Async.promise( args );
					lsRes.push( args.s );
					async.resolve();
				}, "g2" )
			, ( new __.Async() )
				.then( self, "error", {bErr:true,s:"bad1"}, "b1" )
				.then( function( args ) {
					var async = __.Async.promise( args );
					lsRes.push( args.s );
					async.resolve();
				}, "b2" )
		] )
		.then( function( args ) {
			var async = __.Async.promise( args );
			lsRes.push( "catch" );
			async.resolve();

		} )
		.then( function( args ) {
			var async = __.Async.promise( args );
			Test.assert( lsRes, ["catch"], "try catch with parallel tasks wrapping" );
			async.resolve();
		} )
		.start()
	}
}
app3.test();


// branching
// late arrival catch
</script>

</body>
</html>



